public/app/features/plugins/extensions/registry/Registry.ts

```diff
@@ -0,0 +1,46 @@
+import { Observable, ReplaySubject, Subject, firstValueFrom, map, scan, startWith } from 'rxjs';
+
+import { PluginPreloadResult } from '../../pluginPreloader';
+import { deepFreeze } from '../utils';
+
+type ConstructorOptions<T extends Record<string | symbol, unknown>> = {
+  initialState: T;
+};
+
+// This is the base-class used by the separate specific registries.
+export abstract class Registry<T extends Record<string | symbol, unknown>> {
+  private resultSubject: Subject<PluginPreloadResult>;
+  private registrySubject: ReplaySubject<T>;
+
+  constructor(options: ConstructorOptions<T>) {
+    const { initialState } = options;
+    this.resultSubject = new Subject<PluginPreloadResult>();
+    // This is the subject that we expose.
+    // (It will buffer the last value on the stream - the registry - and emit it to new subscribers immediately.)
+    this.registrySubject = new ReplaySubject<T>(1);
+
+    this.resultSubject
+      .pipe(
+        scan(this.mapToRegistry, initialState),
+        // Emit an empty registry to start the stream (it is only going to do it once during construction, and then just passes down the values)
+        startWith(initialState),
+        map((registry) => deepFreeze(registry))
+      )
+      // Emitting the new registry to `this.registrySubject`
+      .subscribe(this.registrySubject);
+  }
+
+  abstract mapToRegistry(registry: T, item: PluginPreloadResult): T;
+
+  register(result: PluginPreloadResult): void {
```

Comment:
Hmm, what do you think about getting rid of `PluginPreloadResult` in this new base Registry, and using something more generic? E.g.:

```ts
register(result: { pluginId: string; configs: K }): void {
```